src/Common/FileSegment.cpp

```diff
@@ -0,0 +1,397 @@
+}
+
+String FileSegment::getOrSetDownloader()
+{
+    std::lock_guard segment_lock(mutex);
+
+    if (downloader_id.empty())
+    {
+        if (download_state != State::EMPTY
+            && download_state != State::PARTIALLY_DOWNLOADED)
+            throw Exception(ErrorCodes::FILE_CACHE_ERROR,
+                            "Can set downloader only for file segment with state EMPTY or PARTIALLY_DOWNLOADED, but got: {}",
+                            download_state);
+
+        downloader_id = getCallerId();
+        LOG_TEST(log, "Set downloader: {}, prev state: {}", downloader_id, stateToString(download_state));
+        download_state = State::DOWNLOADING;
+    }
+    else if (downloader_id == getCallerId())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Attempt to set the same downloader for segment {} for the second time", range().toString());
+
+    LOG_TEST(log, "Returning with downloader: {} and state: {}", downloader_id, stateToString(download_state));
+    return downloader_id;
+}
+
+String FileSegment::getDownloader() const
+{
+    std::lock_guard segment_lock(mutex);
+    return downloader_id;
+}
+
+bool FileSegment::isDownloader() const
+{
+    std::lock_guard segment_lock(mutex);
+    LOG_TEST(log, "Checking for current downloader. Caller: {}, downloader: {}, current state: {}", getCallerId(), downloader_id, stateToString(download_state));
+    return getCallerId() == downloader_id;
+}
+
+FileSegment::RemoteFileReaderPtr FileSegment::getRemoteFileReader()
+{
+    if (!isDownloader())
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR, "Only downloader can use remote filesystem file reader");
+
+    return remote_file_reader;
+}
+
+void FileSegment::setRemoteFileReader(RemoteFileReaderPtr remote_file_reader_)
+{
+    if (!isDownloader())
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR, "Only downloader can use remote filesystem file reader");
+
+    if (remote_file_reader)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Remote file reader already exists");
+
+    remote_file_reader = remote_file_reader_;
+}
+
+void FileSegment::write(const char * from, size_t size)
+{
+    if (!size)
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR, "Writing zero size is not allowed");
+
+    if (availableSize() < size)
+        throw Exception(
+            ErrorCodes::FILE_CACHE_ERROR,
+            "Not enough space is reserved. Available: {}, expected: {}", availableSize(), size);
+
+    if (!isDownloader())
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR,
+                        "Only downloader can do the downloading. (CallerId: {}, DownloaderId: {})",
+                        getCallerId(), downloader_id);
+
+    if (!cache_writer)
+    {
+        auto download_path = cache->path(key(), offset());
+        cache_writer = std::make_unique<WriteBufferFromFile>(download_path);
+    }
+
+    try
+    {
+        cache_writer->write(from, size);
+        cache_writer->next();
```

Comment:
why is calling `finalize`?

Reply:
who*

`finalize` is called in `void FileSegment::completeBatchAndResetDownloader()`